File: Features/Tags/Models/Tag.swift
//
//  Tag.swift
//  ToDo
//
//  Created by Moinuddin Ahmad on 2/9/26.
//

import Foundation
import SwiftData

@Model
final class Tag {
    #Index<Tag>([\.ownerUserID])

    static let defaultTagNames = ["personal", "work", "health", "shopping", "ideas"]

    var cloudID: UUID? = nil
    var ownerUserID: UUID? = nil
    var name: String = ""
    var createdAt: Date = Date()
    var updatedAt: Date? = nil

    var toDos: [ToDo]? = nil
    var nanoDos: [NanoDo]? = nil
    var primaryToDos: [ToDo]? = nil

    init(
        name: String,
        createdAt: Date = Date(),
        updatedAt: Date? = nil,
        toDos: [ToDo] = [],
        nanoDos: [NanoDo] = [],
        cloudID: UUID? = nil,
        ownerUserID: UUID? = nil
    ) {
        self.cloudID = cloudID
        self.ownerUserID = ownerUserID
        self.name = Self.normalizeName(name)
        self.createdAt = createdAt
        self.updatedAt = updatedAt ?? createdAt
        self.toDos = toDos
        self.nanoDos = nanoDos
    }

    var displayName: String {
        let normalized = Self.normalizeName(name)
        return Self.defaultTagNames.contains(normalized) ? String(localized: String.LocalizationValue(normalized)) : normalized
    }

    var allToDos: [ToDo] {
        get { toDos ?? [] }
        set { toDos = newValue }
    }

    var allNanoDos: [NanoDo] {
        get { nanoDos ?? [] }
        set { nanoDos = newValue }
    }

   var linkedTaskCount: Int {
      let uniqueToDoIDs = Set(
         (toDos ?? []).map(\.id) +
         (primaryToDos ?? []).map(\.id)
      )
      let nanoDoCount = nanoDos?.count ?? 0

      return uniqueToDoIDs.count + nanoDoCount
   }

    static func normalizeName(_ value: String) -> String {
        value.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    }

    var syncUpdatedAt: Date {
        updatedAt ?? createdAt
    }

    func markUpdated(_ date: Date = .now) {
        updatedAt = date
    }
}

File: ToDo-Android/app/src/main/java/dev/iamshift/todo/android/core/model/Tag.kt
package dev.iamshift.todo.android.core.model

import java.time.Instant
import java.util.UUID

data class Tag(
    val id: UUID = UUID.randomUUID(),
    val ownerUserId: UUID? = null,
    val name: String,
    val isDefault: Boolean = false,
    val createdAt: Instant = Instant.now(),
    val updatedAt: Instant = createdAt
) {
    init {
        require(name.isNotBlank()) { "Tag name must not be blank" }
        require(name == normalizeName(name)) { "Tag name must be normalized before construction" }
    }

    val displayName: String
        get() = name

    val syncUpdatedAt: Instant
        get() = updatedAt

    fun withName(value: String, at: Instant = Instant.now()): Tag =
        create(
            name = value,
            id = id,
            ownerUserId = ownerUserId,
            isDefault = isDefault,
            createdAt = createdAt,
            updatedAt = at
        )

    companion object {
        const val MAX_NAME_LENGTH = 80
        val defaultTagNames = listOf("personal", "work", "health", "shopping", "ideas")

        fun normalizeName(value: String): String =
            value.trim().lowercase()

        fun create(
            name: String,
            id: UUID = UUID.randomUUID(),
            ownerUserId: UUID? = null,
            isDefault: Boolean = false,
            createdAt: Instant = Instant.now(),
            updatedAt: Instant = createdAt
        ): Tag {
            val normalized = normalizeName(name)
            require(normalized.isNotEmpty()) { "Tag name must not be blank" }
            require(normalized.length <= MAX_NAME_LENGTH) { "Tag name is too long" }
            return Tag(
                id = id,
                ownerUserId = ownerUserId,
                name = normalized,
                isDefault = isDefault,
                createdAt = createdAt,
                updatedAt = updatedAt
            )
        }

        /** Keeps one deterministic tag for each normalized name. */
        fun canonicalTags(tags: Iterable<Tag>): List<Tag> =
            tags.groupBy { normalizeName(it.name) }
                .values
                .mapNotNull { duplicates ->
                    duplicates
                        .sortedWith(
                            compareByDescending<Tag> { it.updatedAt }
                                .thenBy { it.createdAt }
                                .thenBy { it.id.toString() }
                        )
                        .firstOrNull()
                }
                .sortedBy { it.name }
    }
}
